### ISSUE TITLE
[Bug][Qwen3-Omni] Streamed code2wav drops 555 samples at every chunk boundary, causing audible clicks

### ISSUE THREAD

**Issue by edwingao28** [sgl-project/sglang-omni#1354]
## Symptom

Streamed Qwen3-Omni audio has an audible click at chunk boundaries. @RajPatel17  reproduced this and saved the raw PCM before it reached the browser. The waveform shows abrupt jumps: one chunk ends at one audio level and the next chunk begins at a different level.

## Root cause

Each streaming decode in `Code2WavScheduler` runs the vocoder on `[left_context | new_frames]` and trims `context * total_upsample` samples from the left:

https://github.com/sgl-project/sglang-omni/blob/main/sglang_omni/models/qwen3_omni/components/code2wav_scheduler.py#L235-L237

The problem: the vocoder's output is always 555 samples shorter than `window_frames * total_upsample`. The `Qwen3OmniMoeCausalTransConvNet` layers in the HF model trim `kernel - stride` samples on the right at each decoder block. With upsample rates (8, 5, 4, 3) the deficit compounds to 8 → 45 → 184 → 555. This is structural and independent of weights or input length.

In a single full decode this only truncates the very end of the audio, which is harmless. In chunked decode it means **every chunk loses its last 555 samples (~23 ms at 24 kHz)**. The next chunk's emitted audio starts 555 samples later in time, so the stream has a ~23 ms hole at every boundary. With `stream_chunk_size=10` that is a click every 0.8 s of audio.

The upstream `Qwen3OmniMoeCode2Wav.chunked_decode` in transformers has the same flaw, but its default chunk size is 300 frames, so the gap only appears every 6 s and is much less noticeable.

## Evidence

Small random-init model on CPU, comparing full decode against the scheduler's chunking scheme (60 frames, chunk 10, left context 25, `total_upsample=1920`):

```
full decode:    114645 samples
streamed concat: 111870 samples   # 6 chunks x 555 missing = 3330
```

The missing samples are not recoverable from the emitted stream. They fall in the time range `[boundary - 555, boundary)`, which no chunk ever emits.

## Suggested fix

The missing samples do exist in the next chunk's decode window, at offset `context * total_upsample - 555`. So the trim is off by the deficit. Emitting the last `new_frames * total_upsample` samples instead of trimming a fixed left offset fixes both the gap and the length:

```python
wav = wav[..., -(end - start) * self._total_upsample :]
```

Verified on the same setup: streamed output becomes sample-exact in length (114645 == full), and the max error near every boundary drops to ~1e-8 (float noise from truncated attention/conv context). The 555-sample time shift is gone.

Notes for the fix:

- The deficit can be derived from `config.upsample_rates` instead of hardcoding 555, or avoided entirely by the keep-last-N form above.
- The batched path (`run_step`) has the same trim and needs the same change.
- The first chunk (context 0) still ends 555 samples short until the next chunk covers it; the final chunk's tail truncation matches full-decode behavior.

## Repro script

```python
import torch, numpy as np
from transformers.models.qwen3_omni_moe.configuration_qwen3_omni_moe import Qwen3OmniMoeCode2WavConfig
from transformers.models.qwen3_omni_moe.modeling_qwen3_omni_moe import Qwen3OmniMoeCode2Wav

torch.manual_seed(0)
cfg = Qwen3OmniMoeCode2WavConfig(hidden_size=64, num_hidden_layers=2, num_attention_heads=4,
                                 num_key_value_heads=2, intermediate_size=128, decoder_dim=64)
model = Qwen3OmniMoeCode2Wav(cfg).eval()
U = int(model.total_upsample)  # 1920, same as the real checkpoint
T, CHUNK, CTX = 60, 10, 25
codes = torch.randint(0, cfg.codebook_size, (1, cfg.num_quantizers, T))
with torch.no_grad():
    full = model(codes).reshape(-1)
    parts, start = [], 0
    while start < T:
        end = min(start + CHUNK, T)
        ctx = min(CTX, start)
        wav = model(codes[..., start - ctx:end]).reshape(-1)
        parts.append(wav[ctx * U:])          # current scheduler trim
        start = end
    stream = torch.cat(parts)
print(len(full), len(stream))  # 114645 vs 111870
```

**Comment by charliechenye** [comment]
Hi @edwingao28, I’d like to take this issue.

I’ll update both the serial `decode_delta()` path and the batched `run_step()` path to retain the samples corresponding to the captured new frame range, rather than trimming by the nominal context length.

I’ll also add CPU regressions using a deterministic Code2Wav fake with a fixed right-edge output deficit. The tests will cover:

* streamed vs full sample length and alignment;
* recovery of the previous chunk’s missing tail;
* a partial final chunk;
* equivalent behavior in the serial and batched paths; and
* chunks arriving during a batched forward, so the emitted range remains based on the captured window boundaries.

Does that scope match what you had in mind?

**Comment by edwingao28** [comment]
Hey @charliechenye I already raised pr for this issue, thanks for your input!
